custom_components/mammotion/www/agora-client.js

```diff
@@ -530,6 +753,283 @@ class CameraAgoraCard extends HTMLElement {
+        </button>
+        <button id="joystick-left" class="joystick-button joystick-left">
+          <svg viewBox="0 0 24 24" width="24" height="24"><path fill="white" d="M15.41,16.59L10.83,12L15.41,7.41L14,6L8,12L14,18L15.41,16.59Z" /></svg>
+        </button>
+        <button id="joystick-right" class="joystick-button joystick-right">
+          <svg viewBox="0 0 24 24" width="24" height="24"><path fill="white" d="M8.59,16.59L13.17,12L8.59,7.41L10,6L16,12L10,18L8.59,16.59Z" /></svg>
+        </button>
+      `;
+      
+      videoContainer.appendChild(joystickOverlay);
+      
+      // Setup event listeners
+      this._setupJoystickControls();
+    }
+  }
+  
+  // Toggle joystick visibility based on playback state
+  _toggleJoystickVisibility(show) {
+    if (!this._config.enableJoystick) return;
+    
+    const joystickOverlay = this.shadowRoot.getElementById('joystick-overlay');
+    if (joystickOverlay) {
+      if (show) {
+        joystickOverlay.classList.add('visible');
+      } else {
+        joystickOverlay.classList.remove('visible');
+      }
+    }
+  }
+  
+  // Handle joystick button press
+  _handleJoystickPress(direction) {
+    if (!this._disclaimerShown) {
+      this._showMoveDisclaimer(() => {
+        this._disclaimerShown = true;
+        this._sendMoveCommand(direction);
+      });
+    } else {
+      this._sendMoveCommand(direction);
+    }
+  }
+  
+  // Show disclaimer before first movement
+  _showMoveDisclaimer(onAccept) {
+    const videoContainer = this.shadowRoot.getElementById('agora-video');
+    
+    // Create disclaimer overlay
+    const disclaimerOverlay = document.createElement('div');
+    disclaimerOverlay.className = 'disclaimer-overlay';
+    disclaimerOverlay.innerHTML = `
+      <h3>Warning!</h3>
+      <p>By pressing these controls, the robot will physically move.</p>
+      <p>Please be aware that the video feed may not be in real time.</p>
+      <p>Make sure the area around the robot is clear and safe.</p>
+      <button id="disclaimer-accept" class="disclaimer-button">I understand, proceed</button>
+    `;
+    
+    videoContainer.appendChild(disclaimerOverlay);
+    
+    // Add event listener to accept button
+    const acceptButton = disclaimerOverlay.querySelector('#disclaimer-accept');
+    acceptButton.addEventListener('click', (e) => {
+      e.stopPropagation(); // Prevent click from propagating
+      
+      // Stop any active movements before removing the disclaimer
+      Object.keys(this._moveIntervals).forEach(dir => {
+        this._stopContinuousMove(dir);
+      });
+      
+      // Remove the disclaimer after stopping movements
+      videoContainer.removeChild(disclaimerOverlay);
+      
+      // Force mouseup event simulation to ensure any pressed buttons are released
+      document.dispatchEvent(new MouseEvent('mouseup'));
+      document.dispatchEvent(new TouchEvent('touchend'));
```

Comment:
TouchEvent isn't always defined on desktop, this stops the disclaimer from being dismissed properly.